iceoryx_posh/source/roudi/service_registry.cpp

```diff
@@ -17,45 +17,149 @@
 
 #include "iceoryx_posh/internal/roudi/service_registry.hpp"
 
+#include <algorithm>
+
 namespace iox
 {
 namespace roudi
 {
-void ServiceRegistry::add(const capro::IdString_t& service, const capro::IdString_t& instance)
+cxx::expected<ServiceRegistry::Error> ServiceRegistry::add(const capro::ServiceDescription& serviceDescription) noexcept
 {
-    cxx::set::add(m_serviceMap[service].instanceSet, instance);
+    if (!serviceDescription.isValid())
+    {
+        return cxx::error<Error>(Error::SERVICE_DESCRIPTION_INVALID);
+    }
+
+    // Forbid duplicate service descriptions entries
+    for (auto& element : m_serviceDescriptionVector)
+    {
+        if (element == serviceDescription)
+        {
+            return cxx::error<Error>(Error::SERVICE_DESCRIPTION_ALREADY_ADDED);
```

Comment:
I'm not sure if this really should count as error. We have multi publisher support, therefore it is natural that the same service description might be added to the service registry

Reply:
No, it is not an error. But it also should not be stored multiple times but I think you mean just that.

We hence should return success.

Reply:
Right, I didn't meant to store it multiple times. Maye we need an additional counter but that might make it a little brittle. How does the introspection do it? My guess would be we need the same mechanism.

Reply:
AFAIK the introspection was not developed with n:m communication in mind and has the same issue. @FerdinandSpitzschnueffler had a look some time ago, maybe she can enlighten us.

Changed the return to `cxx::success`. My feeling is we'll need a counter in the further course of #415.